src/requirements/requirement-frontend-utils.ts

```diff
@@ -153,6 +153,22 @@ export function getUserRequirements({
   }));
 }
 
+type MatchedRequirementFulfillmentSpecificationBase = {
+  readonly fulfilledBy: 'courses' | 'credits';
+  readonly eligibleCourses: readonly (readonly number[])[];
+  readonly perSlotMinCount: readonly number[];
+  readonly slotNames: readonly string[];
+  readonly minNumberOfSlots?: number;
+};
+
+type MatchedRequirementFulfillmentSpecification =
+  | (MatchedRequirementFulfillmentSpecificationBase & {
+      readonly additionalRequirements?: {
+        readonly [name: string]: MatchedRequirementFulfillmentSpecificationBase;
+      };
+    })
+  | null;
+
```

Comment:
Can we have documentation on what these types are for? Also would these types be good to put in src/requirement-types.d.ts?

Reply:
Documented the types. Don't think it should go to the `requirement-types.d.ts` since these types are usually used by util functions only and not passed as props